### PR TITLE
Indexing Strict Inequalities (Fixed!)

### PR DESCRIPTION
**Pull request by SamArch27**
I thought I'd make a new PR because the old one was getting messy.

@XiaowenHu96 found an issue in my IndexedInequalityTransformer when discharging filter operations, the type was fixed to signed. Once I fixed this the transformation from strict to weak inequalities is correct and passes all the tests.

Transformation 1:  < to <= and != 
Transformation 2:  > to >= and !=

Note that this PR resolves the issue in the master of indexing signed inequalities on the interpreter since now in getAttributesToDischarge(...) there's an explicit check for the interpreter.

### CODE REVIEW

**Comment by XiaowenHu96** [comment]
Just for the record, the transformation from strict inequal to weak inequal is still happening under interpreter mode, but eventually, they won't be indexed if they are unsigned/float.
In theory, we can prevent this transformation for the interpreter at all to save a few dispatches during runtime, but for the sake of simplicity, the current approach is better.

**Comment by SamArch27** [comment]
> Just for the record, the transformation from strict inequal to weak inequal is still happening under interpreter mode, but eventually, they won't be indexed if they are unsigned/float.
> In theory, we can prevent this transformation for the interpreter at all to save a few dispatches during runtime, but for the sake of simplicity, the current approach is better.

Yep, 100% agree. I think in future the RamIndexAnalysis needs to be refactored and then MakeIndex can be constrained further regarding what constraints can be lifted into index operations rather than what is done now where getAttributesToDischarge(...) needs to remove all of the inequalities that can't be indexed.

File: src/ram/transform/MakeIndex.cpp [hunk 1]
```diff
@@ -274,6 +122,55 @@ Own<RamCondition> MakeIndexTransformer::constructPattern(const std::vector<std::
...
     };
 
+    // transform condition list so that every strict inequality becomes a weak inequality + filter
+    // e.g. Tuple[level, element] < <expr> --> Tuple[level, element] <= <expr> and Tuple[level, element] !=
+    // <expr>
+    std::vector<std::unique_ptr<RamCondition>> toAppend;
+    auto it = conditionList.begin();
+    while (it != conditionList.end()) {
+        auto& cond = *it;
+        if (auto* binRelOp = dynamic_cast<RamConstraint*>(cond.get())) {
```
Comment (mmcgr) [inline src/ram/transform/MakeIndex.cpp:132]: Make this an early exit condition:
```
        auto* binRelOp = dynamic_cast<RamConstraint*>(it->get())
        if (binRelOp == nullptr) {
            ++it;
            continue;
        }
```
That makes it slightly more readable, and reduces the need for the big indented block.

Comment (SamArch27) [reply src/ram/transform/MakeIndex.cpp]: Great point.

**Review by mmcgr — APPROVED** [review]
There seems to be some further improvements to make in future PRs, but this seems to be okay to merge for now.

**Pull request merged**